contracts/cw2981-royalties/src/lib.rs

```diff
@@ -0,0 +1,260 @@
+    pub royalty_percentage: Option<u32>,
+    /// The payment address, may be different to or the same
+    /// as the minter addr
+    /// question: how do we validate this?
+    pub royalty_payment_address: Option<String>,
+}
+
+pub type Extension = Option<Metadata>;
+
+pub type MintExtension = Option<Extension>;
+
+pub type Cw2981Contract<'a> = Cw721Contract<'a, Extension, Empty>;
+pub type ExecuteMsg = cw721_base::ExecuteMsg<Extension>;
+
+#[cfg(not(feature = "library"))]
+pub mod entry {
+    use super::*;
+
+    use cosmwasm_std::entry_point;
+    use cosmwasm_std::{Binary, Deps, DepsMut, Env, MessageInfo, Response, StdResult};
+
+    #[entry_point]
+    pub fn instantiate(
+        deps: DepsMut,
+        env: Env,
+        info: MessageInfo,
+        msg: InstantiateMsg,
+    ) -> StdResult<Response> {
+        Cw2981Contract::default().instantiate(deps, env, info, msg)
+    }
+
+    #[entry_point]
+    pub fn execute(
+        deps: DepsMut,
+        env: Env,
+        info: MessageInfo,
+        msg: ExecuteMsg,
+    ) -> Result<Response, ContractError> {
+        Cw2981Contract::default().execute(deps, env, info, msg)
+    }
+
+    #[entry_point]
+    pub fn query(deps: Deps, env: Env, msg: Cw2981QueryMsg) -> StdResult<Binary> {
+        match msg {
+            Cw2981QueryMsg::RoyaltyInfo {
+                token_id,
+                sale_price,
+            } => to_binary(&query_royalties_info(deps, token_id, sale_price)?),
+            Cw2981QueryMsg::CheckRoyalties {} => to_binary(&check_royalties(deps)?),
+            _ => Cw2981Contract::default()
+                .query(deps, env, msg.into())
+                .map_err(|err| err),
+        }
+    }
+}
+
+// NOTE: default behaviour here is to round down
+// EIP2981 specifies that the rounding behaviour is at the discretion of the implementer
+pub fn query_royalties_info(
+    deps: Deps,
+    token_id: String,
+    sale_price: u128,
+) -> StdResult<RoyaltiesInfoResponse> {
+    let contract = Cw2981Contract::default();
+    let token_info = contract.tokens.load(deps.storage, &token_id)?;
+
+    let royalty_percentage = match token_info.extension {
+        Some(ref ext) => match ext.royalty_percentage {
+            Some(percentage) => Percentage::from(percentage),
+            None => Percentage::from(0),
+        },
+        None => Percentage::from(0),
+    };
+    let royalty_from_sale_price = royalty_percentage.apply_to(sale_price);
+
+    let royalty_address = match token_info.extension {
+        Some(ext) => match ext.royalty_payment_address {
+            Some(addr) => addr,
+            None => String::from(""),
+        },
+        None => String::from(""),
+    };
+
+    Ok(RoyaltiesInfoResponse {
+        address: royalty_address,
+        royalty_amount: royalty_from_sale_price,
+    })
+}
+
+pub fn check_royalties(_deps: Deps) -> StdResult<CheckRoyaltiesResponse> {
+    Ok(CheckRoyaltiesResponse {
+        royalty_payments: true,
```

Comment:
Is this actually checking anything?

Reply:
For this contract, as I reasoned it there's nothing to check. This hook is expected to be present to check if the contract does implement CW2981 and signal that on sale royalties should be checked. I think with the implementation at token level it should always return `true` because it's up to the token.

Of course contracts that extend this can determine their own behaviour and replace this function if they have more complex behaviour (for example, you could maintain a secondary index of which tokens actually have royalties, or something).

In this super simple case that isn't necessary however.

This does make me think this explanation should be in the readme, however :joy: as well as perhaps the point about the royalty addr implicitly being able to be an addr/dao/multisig

Reply:
Hmmmmm, interesting. Yeah maybe add a note. :joy: